Incorporate feedback for Patch 9 (Atomic and Locking Routines)

darius-bluespec opened the pull request:
I haven't tested yet, but with the tight schedule I wanted to allow review to start as soon as possible.

darius-bluespec committed:
Fix copyright message.

Adjust year range.  Remove "Contributed By" line.  Replace postal
address with URL.

darius-bluespec committed:
Remove unused atomic_fast types.

darius-bluespec committed:
Remove misleading comment and dead code.

Conditional atomic/!atomic block inside atomic block doesn't make sense.

darius-bluespec committed:
Fix memory ordering for provided atomic macros.

Provide acq and rel varities of atomic_exchange_and_add.  Use acq
ordering for unqualified atomic max, min, and bit_test_set.

darius-bluespec committed:
Add comment explaining Linux atomic helper syscalls.

darius-bluespec committed:
Remove custom pthread lock implementation in favor of generic routine.

darius-bluespec committed:
Simplify __pthread_rwlock_arch_t.

We don't need to preserve binary compatibility (yet).

darius-bluespec committed:
Use compiler built-in functions to implement atomic macros, where possible.

darius-bluespec committed:
Check size in amo macro, 64b ops are not supported on RV32.

darius-bluespec committed:
Combine generic and Linux atomic-machine.h into Linux one.

aswaterman commented on sysdeps/unix/sysv/linux/riscv/atomic-machine.h:
```diff
@@ -33,6 +199,9 @@
 #include <sys/syscall.h>
 #include <sysdep.h>
 
+#define __HAVE_64B_ATOMICS 0
```
Shouldn't this also be ```#define __HAVE_64B_ATOMICS (__riscv_xlen >= 64)```, as with the ```__riscv_atomic``` case above?

aswaterman commented:
Sorry for the delay.  Looks good, but you should take a look at the one comment I had.  I'll leave it to @palmer-dabbelt to handle the merge

darius-bluespec commented:
Zero is definitely wrong, but I think it should be 1, because the
syscall supports 64-bit CAS, even on 32-bit.  Unless maybe we don't want
to use that (on 32-bit) for consistency.


On 06/28/2017 02:13 AM, Andrew Waterman wrote:
> *@aswaterman* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In sysdeps/unix/sysv/linux/riscv/atomic-machine.h
> <https://github.com/riscv/riscv-glibc/pull/14#discussion_r124459110>:
> 
>> @@ -33,6 +199,9 @@
>  #include <sys/syscall.h>
>  #include <sysdep.h>
>  
> +#define __HAVE_64B_ATOMICS 0
> 
> Shouldn't this also be |#define __HAVE_64B_ATOMICS (__riscv_xlen >=
> 64)|, as with the |__riscv_atomic| case above?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/riscv-glibc/pull/14#pullrequestreview-46747474>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn2PewE3O9giMV2XeK-EOd2p4QSIlV_ks5sIe79gaJpZM4OC_OX>.
>

darius-bluespec commented:
I rebased to riscv-for-submission-v1 and addressed the comment by @aswaterman.  I set ```#define __HAVE_64B_ATOMICS (__riscv_xlen >= 64)``` for ```!__riscv_atomic```, because I think that makes the most sense to support going forward, particularly for non-A user space on a kernel that uses A instructions.

aswaterman commented:
Thanks.  @palmer-dabbelt merge at your leisure

palmer-dabbelt commented:
Thanks!
